lib/pc_patch.c

```diff
@@ -660,3 +660,66 @@ pc_patch_set_schema(const PCPATCH *patch, const PCSCHEMA *new_schema)
 
     return (PCPATCH*) paout;
 }
+
+/**
+* Rotate a patch given a unit quaternion.
+*/
+PCPATCH *
+pc_patch_rotate_quaternion(
+    const PCPATCH *patch,
+    double qw, double qx, double qy, double qz,
+    const char *xdimname, const char *ydimname, const char *zdimname)
+{
+    PCPATCH_UNCOMPRESSED *uncompressed_patch;
+    PCDIMENSION *xdim, *ydim, *zdim;
+    const PCSCHEMA *schema;
+    PCPOINTLIST *pointlist;
+    PCMAT33 qmat;
+    PCVEC3 vec, rvec;
+    size_t i;
+
+    pc_matrix_set_from_quaternion(qmat, qw, qx, qy, qz);
+
+    if ( patch->type == PC_NONE )
+    {
```

Comment:
In the case where the input patch is not compressed we could do the rotation in place. But for a consistent API, where the returned patch is alway a new patch, I decided to clone the input patch and do the rotation in that cloned patch.